packages/dx-react-grid/src/plugins/table-column-reordering.jsx

```diff
@@ -0,0 +1,196 @@
+    return draftOrderComputed(order, sourceColumnIndex, targetColumnIndex);
+  }
+  getAvailableColumns() {
+    return this.getDraftOrder()
+      .filter(columnName => !!this.cellDimensionGetters[columnName]);
+  }
+  getCellDimensions() {
+    return this.getAvailableColumns()
+      .map(columnName => this.cellDimensionGetters[columnName]());
+  }
+  storeCellDimensionsGetter(tableColumn, data) {
+    if (tableColumn.type === TABLE_DATA_TYPE) {
+      this.cellDimensionGetters[tableColumn.column.name] = data;
+    }
+  }
+  handleOver({ payload, clientOffset: { x } }) {
+    const sourceColumnName = payload[0].columnName;
+    const availableColumns = this.getAvailableColumns();
+    const relativeSourceColumnIndex = availableColumns.indexOf(sourceColumnName);
+
+    if (relativeSourceColumnIndex === -1) return;
+
+    const cellDimensions = this.getCellDimensions();
+    const overlappedColumns = cellDimensions
+      .filter(({ left, right }) => left <= x && x <= right);
+
+    if (overlappedColumns.length > 1) return;
+
+    const relativeTargetIndex = getTableTargetColumnIndex(
+      cellDimensions,
+      relativeSourceColumnIndex,
+      x,
+    );
+
+    if (relativeTargetIndex === -1) return;
+
+    const {
+      sourceColumnIndex: prevSourceColumnIndex,
+      targetColumnIndex: prevTargetColumnIndex,
+    } = this.getState();
+    const draftOrder = this.getDraftOrder();
+    const targetColumnIndex = draftOrder.indexOf(availableColumns[relativeTargetIndex]);
+
+    if (targetColumnIndex === prevTargetColumnIndex) return;
+
+    const sourceColumnIndex = prevSourceColumnIndex === -1
+      ? draftOrder.indexOf(sourceColumnName)
+      : prevSourceColumnIndex;
+
+    this.setState({
+      sourceColumnIndex,
+      targetColumnIndex,
+    });
+  }
+  handleLeave() {
+    this.setState({
+      sourceColumnIndex: -1,
+      targetColumnIndex: -1,
+    });
+  }
+  handleDrop() {
+    const { sourceColumnIndex, targetColumnIndex, order } = this.getState();
+    const { onOrderChange } = this.props;
+    const nextOrder = changeColumnOrder(order, {
+      sourceColumnName: order[sourceColumnIndex],
+      targetColumnName: order[targetColumnIndex],
+    });
+
+    this.setState({
+      order: nextOrder,
+      sourceColumnIndex: -1,
+      targetColumnIndex: -1,
+    });
+
+    if (onOrderChange) {
+      onOrderChange(nextOrder);
+    }
+  }
+  render() {
+    const { reorderingRowTemplate, reorderingCellTemplate } = this.props;
+    const columnsComputed = ({ tableColumns }) =>
+      orderedColumns(tableColumns, this.getDraftOrder());
+
+    this.cellDimensionGetters = {};
+
+    return (
+      <PluginContainer
+        pluginName="TableColumnReordering"
+        dependencies={pluginDependencies}
+      >
+        <Getter name="tableColumns" computed={columnsComputed} />
+        <Getter name="tableHeaderRows" computed={tableHeaderRowsComputed} />
+        <Template name="tableView">
+          <DropTarget
```

Comment:
It is part of the web components. It should not be present in core plugin. Please, extract an additional template